whylog/constraints/verifier.py

```diff
@@ -0,0 +1,132 @@
+import itertools
+
+from whylog.config import Clue
+from whylog.constraints import IdenticalIntervals, TimeConstraint
+from whylog.constraints.exceptions import UnsupportedConstraintTypeError
+from whylog.front import FrontInput
+
+
+class ConstraintManager(object):
+    @classmethod
+    def get_constraint_by_type(cls, constraint_data):
+        constraints = {
+            'identical': IdenticalIntervals({}),
+            'time': TimeConstraint({})
+            # register your constraint here
+        }
+        if constraint_data['name'] in constraints:
+            return constraints[constraint_data['name']]
+        raise UnsupportedConstraintTypeError(constraint_data)
+
+
+class Verifier(object):
+    UNMATCHED = Clue(None, None, None, None)
+
+    @classmethod
+    def _front_input_from_clue(cls, clue):
+        return FrontInput(clue.line_offset, clue.line_prefix_content, clue.line_source)
+
+    @classmethod
+    def _create_investigation_result(cls, clues_combination, constraints):
+        return InvestigationResult(
+            [cls._front_input_from_clue(clue) for clue in clues_combination], constraints
+        )
+
+    @classmethod
+    def _verify_constraint(cls, combination, effect, constraint):
+        """
+        checks if specified clues (which represents parsers: 1,2,.. for some rule) and
+        effect (which represents parser 0 from this rule) satisfy one given constraint.
+        returns True if so, or False otherwise
+        """
+        constraint_verifier = ConstraintManager.get_constraint_by_type(constraint)
+        groups = []
+        for group_info in constraint['clues_groups']:
+            parser_num, group_num = group_info
+            if parser_num == 0:
+                groups.append(effect.regex_parameters[group_num - 1])
+            else:
+                if combination[parser_num - 1] == Verifier.UNMATCHED:
+                    return False
+                groups.append(combination[parser_num - 1].regex_parameters[group_num - 1])
+        return constraint_verifier.verify(constraint['params'], groups)
+
+    @classmethod
+    def _clues_combinations(cls, clues_tuples, collected_subset=[]):
```

Comment:
Update of this function was caused when one parser occurs many times in rule's causes  list. I think that this function don't work properly. In this implementation (based on your manual example)  we don't get [3, 1, 5], [2, 1, 5] which are necessary to check. Idea is correct, but we must get every pair from [1,2, 3] where order is important in your example